### PR TITLE
fix: Manual float16 conversions to extend platform support

### PR DESCRIPTION
**Pull request by iwoplaza**
React Native (Hermes) unfortunately doesn't yet support Float16Array natively, this PR uses the existing `toHalfBits` and `fromHalfBits` instead.

I also fixed faulty behavior of `toHalfBits` in some edge cases and added tests validation it behaves like Float16Array would.

### CODE REVIEW

File: packages/typegpu/src/data/numeric.ts [hunk 1]
```diff
@@ -198,15 +198,25 @@ export function toHalfBits(x: number): number {
...
-    mant = (mant | 0x800000) >> (1 - exp);
-    mant = (mant + 0x1000) >> 13; // Round‑to‑nearest‑even at bit 10.
-    return (sign << 15) | mant;
+    // Produce a sub‑normal: prepend the hidden 1, then round to nearest,
+    // ties to even. `shift` is the number of low bits dropped from the
+    // 24‑bit significand; the bit just below it is the rounding bit and
+    // everything under that forms the sticky bit.
+    const full = mant | 0x800000; // 24-bit significand incl. the implicit 1.
+    const shift = 14 - exp; // in [14, 24]
+    const roundBit = (full >> (shift - 1)) & 1;
```
Comment (cieplypolar) [inline packages/typegpu/src/data/numeric.ts]:
```suggestion
    const roundBit = (full >>> (shift - 1)) & 1;
```

File: packages/typegpu/src/data/numeric.ts [hunk 2]
```diff
@@ -198,15 +198,25 @@ export function toHalfBits(x: number): number {
...
+    const sticky = full & ((1 << (shift - 1)) - 1) ? 1 : 0;
+    let half = full >>> shift;
+    if (roundBit && (sticky || half & 1)) {
```
Comment (cieplypolar) [inline packages/typegpu/src/data/numeric.ts]:
```suggestion
    if (roundBit & (sticky | half & 1)) {
```

File: packages/typegpu/src/data/numeric.ts [hunk 3]
```diff
@@ -215,18 +225,23 @@ export function toHalfBits(x: number): number {
     return (sign << 15) | 0x7c00; // ±∞
   }
 
-  // 6. Normalised number: round mantissa and pack sign|exp|mant.
-  mant = mant + 0x1000; // Add rounding bias at bit 12.
-  if (mant & 0x800000) {
-    // The carry propagated out of the top bit; mantissa overflowed.
-    mant = 0; // Rounded up to 1.0 × 2^(exp+1).
+  // 6. Normalised number: round mantissa to nearest, ties to even, then pack.
+  const roundBit = (mant >> 12) & 1;
```
Comment (cieplypolar) [inline packages/typegpu/src/data/numeric.ts]:
```suggestion
  const roundBit = (mant >>> 12) & 1;
```

File: packages/typegpu/src/data/numeric.ts [hunk 4]
```diff
@@ -215,18 +225,23 @@ export function toHalfBits(x: number): number {
...
+  const sticky = mant & 0xfff ? 1 : 0;
+  let half = mant >>> 13;
+  if (roundBit && (sticky || half & 1)) {
```
Comment (cieplypolar) [inline packages/typegpu/src/data/numeric.ts]:
```suggestion
  if (roundBit & (sticky | half & 1)) {
```

**Review by cieplypolar — APPROVED** [review]
Very clever bit manipulation 📚 . I left some nits.

**Review by reczkok — APPROVED** [review]
